model-server/src/main/kotlin/org/modelix/model/server/handlers/RepositoriesManager.kt

```diff
@@ -200,18 +207,41 @@ class RepositoriesManager(val client: LocalModelClient) {
             ?: throw IllegalStateException("No version found for branch '${branch.branchName}' in repository '${branch.repositoryId}'")
     }
 
-    private val deltaCache = LRUMap<Pair<String, String?>, Lazy<Map<String, String>>>(10)
-    fun computeDelta(versionHash: String, baseVersionHash: String?): Map<String, String> {
+    private val deltaCache = LRUMap<Pair<String, String?>, SoftReference<Lazy<Map<String, String>>>>(10)
+    fun computeDelta(versionHash: String, baseVersionHash: String?): Flow<Pair<String, String>> {
+        if (versionHash == baseVersionHash) return emptyFlow()
+        if (baseVersionHash == null) {
+            // no need to cache anything if there is no delta computation happening
+            return flow {
+                suspend fun emitObjects(entry: KVEntryReference<*>) {
+                    emit(entry.getHash() to client.get(entry.getHash())!!)
+                    for (referencedEntry in entry.getValue(client.storeCache).getReferencedEntries()) {
+                        emitObjects(referencedEntry)
+                    }
+                }
+
+                emit(versionHash to client.get(versionHash)!!)
+                val version = CLVersion(versionHash, client.storeCache)
+                emitObjects(version.treeHash!!)
+            }
+        }
+
         return runSynchronized(deltaCache) {
-            deltaCache.getOrPut(versionHash to baseVersionHash) {
+            var cachedResult: Lazy<Map<String, String>>?
+            val key = versionHash to baseVersionHash
+            cachedResult = deltaCache.get(key)?.get()
```

Comment:
Somehow GitHub doesn't let me add suggestions... but why not:

```kotlin
val cachedResult = deltaCache.get(key)?.get() ?: {
   ...
}
```

Avoids the mutability and reduces the scope of the variable.